frontend/compute-system/src/components/AppStore/AppRating/AppRating.tsx

```diff
@@ -1,30 +1,144 @@
 import * as React from "react";
-import { Typography, Grid, useMediaQuery } from "@material-ui/core";
+import {
+  Typography,
+  Grid,
+  useMediaQuery,
+  Button,
+  TextField,
+} from "@material-ui/core";
 import ratingStyles from "./appRatingStyles";
 import Stars from "./components/Stars";
+import { DeleteForever, Edit } from "@material-ui/icons";
 
 interface AppRatingProps {
+  id: number;
+  userCommented: boolean;
+  uid: number;
   rate: number;
-  description: string;
+  comment: string;
+  setCommented: (isComment: boolean) => void;
 }
 
 const AppRating = (props: AppRatingProps) => {
   const classes = ratingStyles();
-  const matches = useMediaQuery("(min-width:700px)");
+  const matches = useMediaQuery("(min-width:525px)");
+  const [isEditMode, setEditMode] = React.useState(false);
+  const [commentValid, setValid] = React.useState(true);
+  const [rate, editRate] = React.useState(props.rate);
+  const [rateChanged, setChangedRate] = React.useState(rate);
+  const [comment, editComment] = React.useState(props.comment);
+  const [commentChanged, setChangedComment] = React.useState(comment);
+
+  const changeEditMode = () => {
+    setEditMode(!isEditMode);
+  };
+
+  const deleteComment = () => {
+    props.setCommented(false);
+  };
+
+  const handleComment = (event: React.ChangeEvent<HTMLInputElement>) => {
+    var commentValue = event.target.value;
+    if (commentValue.length > 5000) {
+      setValid(false);
+    } else {
+      setValid(true);
+    }
+    editComment(commentValue);
+  };
+
+  const changeRate = (event: React.ChangeEvent<HTMLInputElement>) => {
+    var value = Number(event.target.value);
+    if (value <= 10 && value >= 0) editRate(value);
+  };
+
+  const cancelChange = () => {
+    setEditMode(!isEditMode);
+    editComment(commentChanged);
+    editRate(rateChanged);
+  };
+
+  const submitChange = () => {
+    setEditMode(false);
+    setChangedComment(comment);
+    setChangedRate(rate);
+  };
 
   return (
     <Grid container alignItems="center" className={classes.grid}>
-      <Grid item container alignItems="center" className={classes.header}>
-        <Grid item xs={matches ? 1 : 2}>
-          <Typography variant="h6" className={classes.rateValue}>
-            {props.rate}/10
-          </Typography>
-        </Grid>
-        <Stars rate={props.rate} />
-      </Grid>
-      <Grid item xs={12} className={classes.comment}>
-        <Typography variant="body1">{props.description}</Typography>
+      <Grid
+        item
+        container
+        alignItems="center"
+        justify={props.userCommented ? "space-between" : undefined}
+        className={matches ? classes.header : classes.headerSmallWindow}
+      >
+        {isEditMode ? (
```

Comment:
can't you merge if from line 76 with if from line 106?

Reply:
if so you can create separate components for each path

Reply:
I don't see an easy way to do that. I will leave it as is